monad-logger-aeson/library/Control/Monad/Logger/Aeson/Internal.hs

```diff
@@ -390,8 +395,18 @@ defaultLogStrLBS now threadContext loc logSource logLevel logStr =
     Text.Encoding.decodeUtf8With Text.Encoding.Error.lenientDecode
       . LBS.toStrict
 
-  logStrLBS = Builder.toLazyByteString logStrBuilder
-  LogStr _ logStrBuilder = logStr
+  logStrLBS = logStrToLBS logStr
+
+logStrToLBS :: LogStr -> LBS.ByteString
+logStrToLBS =
+#if MIN_VERSION_fast_logger(3,0,1)
+  -- Use (presumably) faster/better conversion if we have new enough fast-logger
+  Builder.toLazyByteString . unLogStr
+   where
+    unLogStr (LogStr _ builder) = builder
+#else
+  LBS.fromStrict . fromLogStr
```

Comment:
I tried to trace how this is used. For non-JSON log messages we call `BSL.toStrict` on it, so leaving this strict to begin with shouldn't affect things. In the JSON log messages case, we call `Builder.lazyByteString`, so we could leave this strict and call `Builder.byteString` instead. If we did all that, I don't see any reason to use CPP and would just use `fromLogStr`.

WDYT?

Reply:
Hmm, I'm more comfortable for now keeping the JSON code path as lazy, especially considering the library doesn't yet have benchmarks/throughput tests. For users of `monad-logger-aeson`, it seems likely that the majority of their applications' log messages will be JSON from the outset, making the JSON code path the much more common path (famous last words!). This is appealing, as we only call `toStrict` on the final message (via `defaultLogStrBS`), rather than both on the final message and internally during assembly of the final message (via `fromLogStr`). I feel the way you've used CPP here is the way to go at this time.

I've also had some ideas of changing `defaultOutputOptions`'s signature so that the callback is given a `Builder` instead of a strict `ByteString`. This would let users avoid paying the `toStrict` cost on each log message if their use cases don't require it. At that point, I believe we'd still need CPP like you have here, as there isn't a way to get the `Builder` out with the older versions of `fast-logger`.

Reply:
Fair enough!